docs/doxygen/bda.md

```diff
@@ -0,0 +1,144 @@
+      virtual bool process(BDABuffer) { throw exception; } // Process BDA data.
+    };
+
+With these two process functions, a DPStep can support
+regular and/or BDA data: Existing DPSteps, which only support regular
+data will remain working. Adding support for BDA data can be done
+incrementally, one DPStep at a time.
+
+A disadvantage of this approach is that DPSteps will become larger and
+more complex: They may get two modes of operation, which
+violates the principle of having a single purpose for each class.
+However, if a DPStep becomes too complex, this design allows
+splitting it into a regular variant and a BDA variant for reducing complexity
+
+### BDA data structure
+
+A crucial point in this design is the BDABuffer which holds BDA data.
+The smallest data element in this structure is a measurement row, which
+contains the measurements for all polarisations in all channels for a single
+baseline and a single time step. Since BDA averages data for multiple channels,
+different rows may have different amounts of channels. In code:
+
+    struct Row {
+      double time;
+      double exposure;
+      Matrix<complex> data(nChannels, nPolarisations);
+      Matrix<bool> flags(nChannels, nPolarisations);
+      Matrix<float> weights(nChannels, nPolarisations);
+      Matrix<bool> fullResFlags(nChannels, nPolarisations);
+    };
+
+For reducing overhead, a BDABuffer should not contain a single Row. Using
+a variable-sized vector of rows provides a flexible structure: The BDABuffer
+size can be adjusted to the number of channels in the Rows.
+Combining and splitting buffers is also possible. In code:
+
+    class BDABuffer
+    {
+      vector<Row> itsRows; // Buffer contains a variable number of rows.
+    };
+
+This approach still has a major drawback: Each Matrix object in each row
+performs its own memory allocation, which may yield high overhead.
+A BDABuffer can mitigate this overhead using a shared memory pool for all Rows.
+
+### Steps that use data from multiple iterations.
+
+Steps that use data from multiple iterations have an internal storage
+that holds the data from previous iterations. For regular data, this internal
+storage holds a fixed number of time steps.
+
+For BDA data, the exposure duration of a time step is different for each Row.
+For gathering BDA data for a time interval, an additional BDAIntervalBuffer
+class will be necessary. It should provide the following functionality
+to a step:
+- Set an initial time interval as the current time interval.
+- Store data from BDABuffers.
+- Detect if all relevant data for the current time interval is present.
+- Provide the data for the current time interval. The result should take
+  the BDABuffer start
+  time and exposure duration into account, since they may not match the start
+  time and duration of the requested time interval.
+- Advance the current time interval. The BDAIntervalBuffer should then
```

Comment:
This is all correct. Might make one comment, about this, e.g.: We decided to not implement update functionality. This means we can't immediately implement e.g. DDECal's subtract task for BDA data and apply solutions directly ("applysolutions=true") inside the calibrate task, but during a later restructuring of DP3 we will make streams more generic and implement this in separate subtract and apply tasks that have that functionality."